src/public-lib/lib/persistentApiClientCache.ts

```diff
@@ -0,0 +1,151 @@
+import { set, get, keys, del } from 'idb-keyval';
+import request from 'superagent';
+import internalClient from 'shared/api/cbioportalInternalClientInstance';
+import { getHash } from './apiClientCache';
+
+export type EndpointUpdateInfo = {
+    tables: string[];
+    endpoint: string;
+    timestamp?: Date;
+};
+type TimestampedReponse = { timestamp: Date } & request.Response;
+type Log = (message: string) => void;
+
+export const PERSISTENT_CACHE_PREFIX = 'persistent-cache-'; // prefix to namespace keys from anything else in store
+var endpointUpdateTimes: Map<string, EndpointUpdateInfo> | Promise<void>;
+
+function createCacheCheckingVersionOfGet(
+    name: string,
+    api: any,
+    log?: Log
+): void {
+    const oldMethod = api[name];
+    const method = async function(parameters: any): Promise<request.Response> {
+        // Wait for timestamps to finish loading to avoid race condition
+        // where the persistent cache would sometimes be ignored
+        if (endpointUpdateTimes instanceof Promise) {
+            await endpointUpdateTimes;
+        }
+
+        // for every GET in the api, there is a corresponding
+        // GETURL function that returns the url as a string
+        if (!name.endsWith('WithHttpInfo')) {
+            if (log) {
+                log(
+                    "Attempt to cache method that does not end in 'WithHttpInfo'"
+                );
+                log('Method: ' + name);
+                log('This method will not be cached.');
+            }
+            return await oldMethod.apply(this, [parameters]);
+        }
+        const url = name + '?' + getHash(parameters);
+        // check the cache for a matching url
+        var response: TimestampedReponse = await get(
+            PERSISTENT_CACHE_PREFIX + url
+        );
+
+        // if not in cache, fetch a response and cache it
```

Comment:
we are most likely ok here, but there is a subtle difference between this and the other cache we use: the other cache returns the actual promise that is returned by the "oldMethod."  That makes me more confident that it's a silent "spy" on the original method being proxied.   here we are returning a new promise which will resolve with a response.  i would just want to make double sure that errors errors are still caught properly by calling code.

Reply:
Reminder to self: still haven't done this. Do tomorrow.